doc/developer/design/20230802_webhook_idempotency.md

```diff
@@ -0,0 +1,403 @@
+envelope state and a simple means of doing that is comparing two calls to a
+`NowFn` and determining if their difference is within some configurable bound.
+
+To simplify this tracking, we can include one of the `NowFn` calls as a column
+of data, which we can then compare to another call to a `NowFn` when assessing
+which data to keep in the envelope state. While this is remedial, it's something
+simple that we can generalize in some way.
+
+More on this in the **Upsert state reduction** section of the design doc.
+
+Note that because we want to bound the amount of state we keep in the upsert
+envelope, we must have _something_ that correlates time to a write. An
+alternative here would be to include batch numbers in the row in lieu of a
+timestamp, and change the compaction window of the remap collection to whatever
+retention period we want--however, this adds complexity in successfully dropping
+the data from the upsert envelope.
+
+#### Idempotency guarantees
+
+We plan to offer `UPSERT` idempotency, which means that we will prevent adding
+state to a persist shard if we receive a second write whose `kv` pair matches
+one we have in the `UPSERT` state.
+
+However, this means if we receive multiple requests with the same idempotency
+token, we will retract the initial state and leave the last-seen value in the
+persist shard. Because we plan have add the time the request was received to the
+row of data we append, this means we will continually update the time the
+request was written and extend the lifetime of the idempotency key from the last
+time the write occurred.
+
+I think this is reasonable, but wanted to make sure this behavior's is
+explicitly understood as a tradeoff of storing the time.
+
+**Alternative**
+
+It's possible to achieve a similar outcome by increasing the compaction period
+for the remap shard such that we can get the time at which each event occurred
+in Materialize time--this feedback loop seems challenging to instrument,
+however.
+
+### Architectural overview
+
+#### Dataflow
+
+##### Source rendering
+
+The rendered dataflow will simply read updates out of a channel and give them to
+a data capability at some cadence, downgrading on either some interval's tick or
+when the amount of buffered data reaches some threshold.
+
+Providing users a threshold of requests to buffer offers some knob to control
+throughput vs. latency for the source.
+
+**Timestamping**
+
+The native timestamp of the source could just be a `u64`, which we increment for
+each batch we produce.
+
+The annoying corner of this, though, is that the latency of any request is
+limited by the frequency with which we reclock batches. The most feasible option
+I see here is providing users the option to reclock on a shorter period. The
+current period is 1s, but it could be set to some configurable duration. I'm not
+sure how low this could be set to produce a meaningful effect for the user.
```

Comment:
https://github.com/MaterializeInc/materialize/issues/19322 is tracking the re-enabling of the `TIMESTAMP INTERVAL` option.